### ISSUE TITLE
[WP1][Enhancement] Major refactoring of the Simplified HRM (L-Module Only)

### ISSUE THREAD

**Issue by KyryloKozlovskyi** [fionntmcc/cross-platform-hrm#57]
# Refactor: Rename ge2025 → simplified_hrm, split model into modules, fix training script, add inference scripts

## Summary

Major refactoring of the Simplified HRM (L-Module Only) codebase covering four areas:

1. **Model split** — Broke monolithic `model_ge2025.py` (704 lines) into modular layer files
2. **Rename** — All `ge2025` / `LModuleOnly` naming → `simplified_hrm` / `SimplifiedHRM` 
3. **Training script fixes** — 8 paper-alignment corrections + dead code removal + mixed difficulty support
4. **Inference scripts** — New `run_simplified.py`, `demo_simplified.py`, `visualise_simplified.py`

---

## 1. Model Split into Modules

Split the monolithic `model_ge2025.py` into focused layer files following existing repo conventions (`input_network.py`, `output_network.py`, etc.):

### New Files

| File | Contains | Lines |
|---|---|---|
| `hrm/layers/input_simplified.py` | `InputEmbedding` — Token + puzzle-type embedding | ~110 |
| `hrm/layers/output_simplified.py` | `OutputHead` — Puzzle-specific LM heads (4x4/9x9/maze) | ~100 |
| `hrm/model_simplified.py` | `SimplifiedHRM`, `SimplifiedHRMConfig`, factories, convenience methods | ~450 |

### Design Decisions

- **Circular import avoidance**: Layer files use `TYPE_CHECKING` guards — they only import `SimplifiedHRMConfig` and `PuzzleType` for type hints at static analysis time, not at runtime
- **Backward compatibility**: `model_simplified.py` exports aliases (`LModuleOnlyHRM = SimplifiedHRM`, `create_lmodule_only_hrm = create_simplified_hrm`, etc.) so existing checkpoints and code keep working
- **Zero training script changes needed**: All imports resolve through the main model file

---

## 2. Rename: ge2025 → simplified_hrm

Consistent naming across the entire codebase:

| Old Name | New Name |
|---|---|
| `model_ge2025.py` | `model_simplified.py` |
| `input_ge2025.py` | `input_simplified.py` |
| `output_ge2025.py` | `output_simplified.py` |
| `train_ge2025.py` | `train_simplified.py` |
| `LModuleOnlyHRM` | `SimplifiedHRM` |
| `LModuleOnlyConfig` | `SimplifiedHRMConfig` |
| `create_lmodule_only_hrm()` | `create_simplified_hrm()` |
| `create_small_lmodule_hrm()` | `create_small_simplified_hrm()` |
| `ge2025_hrm_{puzzle}_best.pt` | `simplified_hrm_{puzzle}_best.pt` |
| `training_history_ge2025_` | `training_history_simplified_` |

Updated `hrm/__init__.py` and `hrm/layers/__init__.py` with all new exports.

---

### Paper-Alignment Fixes (Ge et al. 2025)

| # | Issue | Old Value | Fixed Value | Paper Reference |
|---|---|---|---|---|
| 1 | Label smoothing default | `0.1` | `0.0` | Not used in paper |
| 2 | Weight decay hardcoded, no CLI arg | `0.05` | `--weight-decay` arg, default `0.1` | Table 1 |
| 3 | Learning rate default | `1e-4` | `3e-4` | Table 1 |
| 4 | Num samples default | `10000` | `1000` | Section 4.1 |
| 5 | Eval routed through multistep loss | Deep supervision | Final-step only always | Section 3.2 |
| 6 | Cosine annealing min LR | `lr * 0.01` | `lr * 0.1` | lr_min_ratio=0.1 |

### Dead Code Removed

- `compute_multistep_loss()` function (paper doesn't use deep supervision)
- `--multistep-loss` / `use_multistep_loss` CLI arg and logic
- `total_deep_loss` tracking + `output['deep_supervision_loss']` check (model never outputs this key)
- Unused imports: `import torch.nn as nn`, `from datetime import datetime`

### Features Added

- `--data-output-path` CLI arg for custom data save location
- `--difficulty mixed` option — cycles through easy/medium/hard for diverse training data
  - 9x9: cycles through `Difficulty.EASY/MEDIUM/HARD`
  - 4x4: cycles through clue counts `10/8/6`

### Minor Fix

- Fixed invalid escape sequence `\&` in docstring (line 8) — should be `&`

---

## 4. New Inference Scripts

Three new scripts that import `SimplifiedHRM` (not the old `UnifiedHRM`):

### `scripts/run_simplified.py` — Evaluation

Loads a checkpoint, evaluates on dataset, displays colour-coded examples with accuracy stats.

```bash
python scripts/run_simplified.py --model model/simplified_hrm_sudoku_4x4_final.pt --puzzle sudoku_4x4
python scripts/run_simplified.py --model model/simplified_hrm_sudoku_4x4_final.pt --puzzle sudoku_4x4 --data data/sudoku_4x4_train.npz
```

Features: batch evaluation with token/puzzle accuracy metrics, colour-coded terminal output (bold=given, green=correct, red=wrong), auto-detects data files.

### `scripts/demo_simplified.py` — Quick Demo & Interactive Mode

Hardcoded example puzzles + interactive mode for manual puzzle entry.

```bash
python scripts/demo_simplified.py --model model/simplified_hrm_sudoku_4x4_final.pt --puzzle sudoku_4x4
python scripts/demo_simplified.py --model model/simplified_hrm_sudoku_4x4_final.pt --interactive
```

### `scripts/visualise_simplified.py` — Reasoning Step Visualisation

Shows the model's predicted grid after each of the N reasoning steps, demonstrating the iterative refinement / diffusion-like denoising process.

```bash
python scripts/visualise_simplified.py --model model/simplified_hrm_sudoku_4x4_final.pt --puzzle sudoku_4x4
python scripts/visualise_simplified.py --model model/simplified_hrm_sudoku_4x4_final.pt --puzzle sudoku_4x4 --delay 1.0
```

Features: step-by-step grid display, tracks cells changed between steps, accuracy at each step, configurable delay for animation effect.

---

## 5. Files Safe to Move to `prototype/`

The following files are no longer in the active dependency chain for SimplifiedHRM training/inference:

| File | What it is |
|---|---|
| `hrm/model.py` | Original full HRM (H+L, MLP-based) |
| `hrm/model_simple.py` | Earlier simplified attempt (single-cell prediction) |
| `hrm/model_unified.py` | Full UnifiedHRM with ACT/Q-learning |
| `hrm/model_ge2025.py` | Old monolithic file (replaced by model_simplified.py) |
| `hrm/layers/input_network.py` | MLP-based InputNetwork |
| `hrm/layers/output_network.py` | Cell+digit OutputNetwork |
| `hrm/layers/worker.py` | WorkerModule (full HRM inner loop) |
| `hrm/layers/planner.py` | PlannerModule / H-module |
| `hrm/core/iteration.py` | Fixed-point iteration |
| `hrm/core/iteration_outer.py` | Hierarchical convergence |
| `hrm/core/halting.py` | ACT / Q-learning halting |
| `hrm/data/dataset.py` | Single-cell SudokuDataset |
| `scripts/train_unified.py` | Training script for UnifiedHRM |
| `scripts/run_model.py` | Evaluation script for UnifiedHRM |
| `scripts/demo_unified.py` | Demo script for UnifiedHRM |
| `scripts/visualise_steps.py` | Visualisation for UnifiedHRM |

After moving, `hrm/layers/__init__.py` and `hrm/__init__.py` can be slimmed to only export active components.

---

### New
- `hrm/layers/input_simplified.py`
- `hrm/layers/output_simplified.py`
- `scripts/run_simplified.py`
- `scripts/demo_simplified.py`
- `scripts/visualise_simplified.py`

### Modified
- `hrm/model_simplified.py` (slimmed — layers extracted, renamed from model_ge2025.py)
- `hrm/layers/__init__.py` (added new exports)
- `hrm/__init__.py` (added new exports + backward-compat aliases)
- `scripts/train_simplified.py` (8 fixes + dead code removal + mixed difficulty + rename)

### Can Be Removed / Moved to prototype
- `hrm/model_ge2025.py` (replaced by model_simplified.py)
- All files listed in section 5 above

---

## Training Commands (Post-Refactor)

```bash
# Run 1: Full 9x9 (overnight, ~15-20hrs on RTX 3060)
python scripts/train_simplified.py --puzzle sudoku_9x9 --generate-data --num-samples 50000 --difficulty mixed --epochs 150 --batch-size 64 --lr 3e-4 --weight-decay 0.1 --amp

# Run 2: Small 4x4 (~25 min)
python scripts/train_simplified.py --puzzle sudoku_4x4 --small --generate-data --num-samples 5000 --difficulty mixed --epochs 100 --batch-size 64 --lr 3e-4 --weight-decay 0.1

# Run 3: Small 9x9 (~7hrs)
python scripts/train_simplified.py --puzzle sudoku_9x9 --small --generate-data --num-samples 50000 --difficulty mixed --epochs 200 --batch-size 64 --lr 3e-4 --weight-decay 0.1 --amp
```

---

## Verification

- All files parse (AST validated)
- No stale `ge2025` references outside paper citations
- Backward-compatible aliases ensure old checkpoint loading works
- Training script needs zero import changes (all resolve through `model_simplified.py`)
- 21 automated checks passed on the model split
